---
title: "DATA3888 Biotechnology Project - Generate Images"
author: "Semester 1, 2024"
params:
  soln: TRUE   ## EDIT to TRUE when generating solution, otherwise 'FALSE'
  supp: FALSE
  show: 'as.is'  ## EDIT to 'as.is' when generating Suggestions, otherwise 'hide'
output:
  html_document:
    fig_caption: yes
    include:
      after_body: css/stylesDD.js
    number_sections: yes
    embed-resources: true
    theme: flatly
    css: 
      - css/styles.css
      - https://use.fontawesome.com/releases/v5.0.6/css/all.css
    toc: true
    toc_depth: 3
    toc_float: true
    code_folding: hide
  pdf_document:
    number_sections: yes
    toc: yes
editor_options: 
  markdown: 
    wrap: 72
---

```{r initialsetup, include=FALSE}
# For suggestions
chooseCRANmirror(graphics=FALSE, ind=1)
knitr::opts_chunk$set(echo = TRUE,tidy=TRUE)
show_q = FALSE
show_s = FALSE
# Then include suggestions file
# ```{r child='Lab1_Quiz_s.Rmd', eval = show_s}
```

<br><br>

::: aimbox
### [<i class="fa fa-circle fa-stack-2x"></i><i class="fa fa-pencil-alt fa-stack-1x fa-inverse"></i>]{.fa-stack .fa} Document summary {.unnumbered}

-   This RMarkdown document serves as a reproducible code for generating
    the Biotechnology per-cell images.
-   This can be fully *re-run* assuming:
    -   Enough disk capacity to download full Biotechnology data bundle,
    -   Enough memory to read `ome.tif` image via `RBioFormats` package,
    -   Similar directory structure as described below, and
    -   Mac/linux operating system.
-   The *output* is the zipped file `Biotechnology.zip`, which has
    already been provided on Canvas.

### [<i class="fa fa-circle fa-stack-2x"></i><i class="fa fa-location-arrow fa-stack-1x fa-inverse"></i>]{.fa-stack .fa} Key considerations {.unnumbered}

-   Only a randomly selected subset of 1,000 cells' images are extracted
    in this document.
-   Cell labels are the result of Graph-based clustering (an
    *unsupervised* learning technique) of *gene expression* data
    extracted for these cells. Can this be thought of as "ground truth"?
    <br>
:::

<br>

# Full data description

The data described in this document stems from a new biotechnology:
molecule-resolved spatial genomics. In particular we will explore data
that has been generated by [10x Genomics](https://www.10xgenomics.com/)
Xenium instrument on a fresh frozen mouse brain coronal section - Tiny
subset. The technology results in several outputs including:

-   cell morphology image where intensity corresponds to presence of the
    nucleus of each cell.
-   cell boundaries indicating spatial locations of detected cells
-   RNA abundances (gene expression) of each cell, which has been
    grouped into 28 distinct clusters, of which cluster labels are
    provided.

The full data and description can be found in this
[link](https://www.10xgenomics.com/resources/datasets/fresh-frozen-mouse-brain-for-xenium-explorer-demo-1-standard).

# Data and code directory structure

For this reproducible code we assume that this RMarkdown document is
saved within the following directory structure:

-   `Biotechnology/`
    -   `data_processed/`
        -   `clusters.csv`
        -   `cell_boundaries.csv.gz`
        -   `morphology_focus.tif`
    -   `data_raw/`
        -   `Xenium_V1_FF_Mouse_Brain_Coronal_Subset_CTX_HP_outs.zip`
        -   `<unzipped files>`
    -   `scripts/`
        -   `DATA3888_Biotechnology_generateImages_2024.Rmd` (this
            document)

For you to be able to fully re-run this code you will need to download
the contents of `data_raw/` separately (see next section).

You are provided this directory structure, with the contents of
`data_raw/` removed due to the large file size.

# Raw data bundle

The contents to `Xenium_V1_FF_Mouse_Brain_Coronal_Subset_CTX_HP_outs`
folder is from the
`Xenium_V1_FF_Mouse_Brain_Coronal_Subset_CTX_HP_outs.zip` file (approx
3.5GB), available to download via this
[LINK](https://cf.10xgenomics.com/samples/xenium/1.0.2/Xenium_V1_FF_Mouse_Brain_Coronal_Subset_CTX_HP/Xenium_V1_FF_Mouse_Brain_Coronal_Subset_CTX_HP_outs.zip),
or can be programmatically downloaded using `wget` into the target
directory.

```         
wget https://cf.10xgenomics.com/samples/xenium/1.0.2/Xenium_V1_FF_Mouse_Brain_Coronal_Subset_CTX_HP/Xenium_V1_FF_Mouse_Brain_Coronal_Subset_CTX_HP_outs.zip ../data_raw/
unzip ../data_raw/Xenium_V1_FF_Mouse_Brain_Coronal_Subset_CTX_HP_outs.zip -d ../data_raw/
```

Note! It is very important to ensure you are working from the *correct
working directory*, i.e. within the `scripts` folder in the directory
structure described above.

# The `EBImage` package

`EBImage` is an R package that is available in the [Bioconductor
Project](http://bioconductor.org/). Bioconductor is similar to the
[Comprehensive R Archive Network (CRAN)](https://cran.r-project.org/),
in that you can install packages from this repository.

[`EBImage`](https://bioconductor.org/packages/release/bioc/html/EBImage.html)
provides general purpose functionality for image processing and
analysis. In the context of (high-throughput) microscopy-based cellular
assays, EBImage offers tools to segment cells and extract quantitative
cellular descriptors. This allows the automation of such tasks using the
R programming language and facilitates the use of other tools in the R
environment for signal processing, statistical modeling, machine
learning and visualization with image data.

This [chapter in Modern Statistics for Modern
Biology](https://web.stanford.edu/class/bios221/book/11-chap.html) is a
great reference for using `EBImage` for different types of imaging data.

To install the `EBImage` package, you can run the chunk below. This will
check whether you have the `BiocManager` package installed, and if not
it will install `BiocManager`. Then, the `EBImage` package will be
installed via the `BiocManager::install()` function.

Note: if you attempt to run `install.packages("EBImage")` you may be met
with an error! This is because the package is available in Bioconductor
and not on CRAN.

```         
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("EBImage")
```

# Convert cell morphology `.ome.tif` to `.tif` format

The raw data bundle contains the cell morphology image in the
[`ome.tif`](https://docs.openmicroscopy.org/ome-model/5.6.3/ome-tiff/)
file format. This type of file includes the image pixel intensities as
well as additional metadata that is associated with the microscopy
experiment. Since we are interested in the image information only, we
need to convert to a `.tif` format to enable further downstream
processing with `EBImage`.

You are given the `.tif` file in the Processed Data Bundle, but you can
see how this was generated in the rest of this section.

Load the `EBImage` R package.

```{r, setup}
library(EBImage)
```

Read in morphology focus `.ome.tif` image and export out as a `.tif`
into the `../data/` folder. Only do so if the target `.tif` file does
not exist.

Note that if we need to generate the `.tif` file, we need to first set
up the java memory to 10GB and load the `RBioFormats` package, which is
available on [Bioconductor development
branch](https://www.bioconductor.org/packages/devel/bioc/html/RBioFormats.html)
and on [Github](https://github.com/aoles/RBioFormats).

```{r}
tifFile = "../data_processed/morphology_focus.tif"
if (!file.exists(tifFile)) {
  options(java.parameters = "-Xmx10g")
  library(RBioFormats)
  checkJavaMemory()
  img_ome = RBioFormats::read.image("../data_raw/morphology_focus.ome.tif",
                                    read.metadata = FALSE,
                                    normalize = TRUE)
  
  img = img_ome[[1]]@.Data
  EBImage::writeImage(x = img,
                      files = tifFile,
                      type = "tiff")
}
```

# Copy tabular cell data to `../data_processed/` directory

Since the raw data bundle contains many large files, for convenience we
have copied two files from the `../data_raw/` directory to the
`../data_processed/` directory. This can be done programmatically using
the `system()` function.

```         
system("cp ../data_raw/cell_boundaries.csv.gz ../data_processed/cell_boundaries.csv.gz")
system("cp ../data_raw/analysis/clustering/gene_expression_graphclust/clusters.csv ../data_processed/clusters.csv")
```

# Read cell morphology image data

Read and display the morphology image. Display requires some scaling of
the intensities according to the distribution of the intensities, to the
99th percentile.

```{r}
img = EBImage::readImage(tifFile)
EBImage::display(img/quantile(img,0.99))
```

# Read cell segmentation data

Cell segmentation is provided in the data bundle as a `.csv` file
containing the vertices around each cells' boundary. The coordinates of
the boundaries need to be converted between micrometres (um) and pixels.
This scaling factor can be found in the `../data_raw/experiment.xenium`
file under "pixel_size".

Note that `read.csv` can read a
[Gzip](https://en.wikipedia.org/wiki/Gzip) compressed file.

```{r}
cell_boundaries = read.csv("../data_processed/cell_boundaries.csv.gz", header = TRUE)
cell_boundaries$vertex_x_trans = cell_boundaries$vertex_x/0.2125
cell_boundaries$vertex_y_trans = cell_boundaries$vertex_y/0.2125
head(cell_boundaries)
```

# Read gene expression-based cluster labels for all cells

The imaging data also contains gene expression information, that has
been used to perform graph-based clustering. We read this data in via
`read.csv`.

```{r}
clusters = read.csv("../data_processed/clusters.csv")
head(clusters)
ncells = nrow(clusters)
ncells
```

# Generate per-cell images for 1,000 randomly selected cells

In this code chunk, we extract morphology images for 1,000 random cells.
For each cell, we subset the morphology image to the rectangle of pixels
that cover the cell segmentation boundary.

To get a sense of the variety of cell morphologies, we visualise the
first five randomly selected cells.

```{r}
set.seed(2024)

ncells_subset = 1000

cells_subset = sample(ncells,ncells_subset)
table(clusters[cells_subset,"Cluster"], useNA = "always")

for (i in cells_subset) {
  
  # extract the boundary vertices for the selected cell
  bounds_i = subset(cell_boundaries,cell_id == i)
  
  # extract the cluster value for the selected cell
  clustval_i = clusters[i,"Cluster"]
  
  # extract the pixel intensities for the area covering the cell boundary
  img_sub = img[min(bounds_i$vertex_x_trans):max(bounds_i$vertex_x_trans),
                min(bounds_i$vertex_y_trans):max(bounds_i$vertex_y_trans)]
  
  # normalise the pixel intensities according to 99th percentile
  img_sub_norm = img_sub/quantile(img_sub, 0.99)
  
  # as an example, display the image for the first selected cell
  if (i %in% cells_subset[1:5]) {
    print(paste0("displaying image for cell ", i))
    EBImage::display(img_sub/quantile(img_sub, 0.99))
  }
  
  # if there is no folder for cell_images, create one
  if (!file.exists("../data_processed/cell_images/")) {
    system("mkdir ../data_processed/cell_images/")
  }
  
  # if there is no folder for the cluster, create one
  clustval_i_directory = paste0("../data_processed/cell_images/cluster_",clustval_i)
  if (!file.exists(clustval_i_directory)) {
    system(paste0("mkdir ", clustval_i_directory))
  }
  
  # save the extracted image as a png file
  EBImage::writeImage(x = img_sub_norm,
                      files = paste0(clustval_i_directory, "/cell_", i, ".png"),
                      type = "png")
  
}
```

# Create `cell_images.zip` processed data file

The contents of the `data_processed/cell_images` folder can then zipped
into a file to be shared separately, with the following commands in the
terminal. The first command changes the working directory to
`../data_processed/` and the next command creates the `cell_images.zip`
file, containing all the contents of the `cell_images/` folder.

```         
cd ../data_processed/
zip -r cell_images.zip cell_images/*
```

# Finish

```{r}
sessionInfo()
```

```{r}
knitr::knit_exit()
```
